src/Type/ArrayType.php

```diff
@@ -356,9 +356,29 @@ public function setOffsetValueType(?Type $offsetType, Type $valueType, bool $uni
 
 	public function setExistingOffsetValueType(Type $offsetType, Type $valueType): Type
 	{
-		return new self(
-			$this->keyType,
-			TypeCombinator::union($this->itemType, $valueType),
+		if ($offsetType instanceof ConstantStringType || $offsetType instanceof ConstantIntegerType) {
```

Comment:
The idea is that we're overwriting an existing offset so `$this->keyType` should stay intact as is in the original version of this method.

Of course setExistingOffsetValueType was used when unsetting an offset so some details somewhere might be wrong, but what you're doing here now is basically the same thing `setOffsetValueType` is doing so it seems wrong.